brave/src/main/java/brave/internal/baggage/UnsafeArrayMap.java

```diff
@@ -0,0 +1,388 @@
+      Map.Entry that = (Map.Entry) o;
+      int i = arrayIndexOfKey(that.getKey());
+      if (i == -1) return false;
+      return value(i + 1).equals(that.getValue());
+    }
+  }
+
+  abstract class SetView<E> implements Set<E> {
+    int advancePastFiltered(int i) {
+      while (i < toIndex && isFilteredKey(filteredBitSet, i)) i += 2;
+      return i;
+    }
+
+    @Override public int size() {
+      return size;
+    }
+
+    /**
+     * By abstracting this, {@link #keySet()} {@link #values()} and {@link #entrySet()} only
+     * implement need implement two methods based on {@link #<E>}: this method and and {@link
+     * #contains(Object)}.
+     */
+    abstract E elementAtArrayIndex(int i);
+
+    @Override public Iterator<E> iterator() {
+      return new ReadOnlyIterator();
+    }
+
+    @Override public Object[] toArray() {
+      return copyTo(new Object[size]);
+    }
+
+    @Override public <T> T[] toArray(T[] a) {
+      T[] result = a.length >= size ? a
+          : (T[]) Array.newInstance(a.getClass().getComponentType(), size());
+      return copyTo(result);
+    }
+
+    <T> T[] copyTo(T[] dest) {
+      for (int i = 0, d = 0; i < toIndex; i += 2) {
+        if (isFilteredKey(filteredBitSet, i)) continue;
+        dest[d++] = (T) elementAtArrayIndex(i);
+      }
+      return dest;
+    }
+
+    final class ReadOnlyIterator implements Iterator<E> {
+      int i = advancePastFiltered(0);
+
+      @Override public boolean hasNext() {
+        i = advancePastFiltered(i);
+        return i < toIndex;
+      }
+
+      @Override public E next() {
+        if (!hasNext()) throw new NoSuchElementException();
+        E result = elementAtArrayIndex(i);
+        i += 2;
+        return result;
+      }
+
+      @Override public void remove() {
+        throw new UnsupportedOperationException();
+      }
+    }
+
+    @Override public boolean containsAll(Collection<?> c) {
+      if (c == null) return false;
+      if (c.isEmpty()) return true;
+
+      for (Object element : c) {
+        if (!contains(element)) return false;
+      }
+      return true;
+    }
+
+    @Override public boolean isEmpty() {
+      return false;
+    }
+
+    @Override public boolean add(E e) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override public boolean remove(Object o) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override public boolean addAll(Collection<? extends E> c) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override public boolean retainAll(Collection<?> c) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override public boolean removeAll(Collection<?> c) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override public void clear() {
+      throw new UnsupportedOperationException();
+    }
+  }
+
+  static final class Entry<K, V> implements Map.Entry<K, V> {
```

Comment:
Can we use `SimpleImmutableEntry` instead of this class?

Reply:
sure can. I didn't know that was public! also it is the same as are lowest java level. score!

Reply:
80 less lines thanks to you!